subsys/fs/nvs/nvs.c

```diff
@@ -781,10 +781,21 @@ static int nvs_startup(struct nvs_fs *fs)
 			}
 		}
 	}
-	/* all sectors are closed, this is not a nvs fs */
+	/* all sectors are closed, this is not a nvs fs or irreparably corrupted */
 	if (closed_sectors == fs->sector_count) {
+#ifdef CONFIG_NVS_BAD_PARTITION_RECOVERY
+		LOG_WRN("All sectors closed, erasing all settings...");
+		addr = 0U;
+		rc = flash_erase(fs->flash_device, 0, fs->sector_size * fs->sector_count);
+		if (rc) {
+			rc = -EIO;
+			goto end;
+		}
+		i = 0;
```

Comment:
I think we want:
`i = fs->sector_count;` 
`addr = ((fs->sector_count - 1) << ADDR_SECT_SHIFT) + (uint16_t)(fs->sector_size - ate_size);`